Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -0,0 +1,355 @@
+
+import * as fs from 'fs';
+
+import { xhr, getErrorStatusDescription } from 'request-light';
+import URI from 'vscode-uri';
+import { IConnection, TextDocuments } from 'vscode-languageserver';
+import {
+  TextDocument,
+  Diagnostic,
+  CompletionList,
+  Hover,
+  Position,
+  CompletionItemKind,
+  CompletionItem,
+  Range,
+} from 'vscode-languageserver-types';
+import { TextDocumentPositionParams } from 'vscode-languageserver-protocol';
+import * as lg from 'botbuilder-lg';
+import get from 'lodash/get';
+import { LGTemplate } from 'botbuilder-lg';
+
+import { buildInfunctionsMap } from './builtinFunctionsMap';
+import {
+  getRangeAtPosition,
+  convertSeverity,
+  getLGResources,
+  updateTemplateInContent,
+  getTemplatePositionOffset,
+  LGDocument,
+} from './utils';
+
+// define init methods call from client
+const InitializeDocumentsMethodName = 'initializeDocuments';
+
+const allowedCompletionStates = ['plaintext', 'expression'];
+
+export class LGServer {
+  protected workspaceRoot: URI | undefined;
+  protected readonly documents = new TextDocuments();
+  protected readonly pendingValidationRequests = new Map<string, number>();
+  protected LGDocuments: LGDocument[] = []; // LG Documents Store
+
+  constructor(protected readonly connection: IConnection) {
+    this.documents.listen(this.connection);
+    this.documents.onDidChangeContent(change => this.validate(change.document));
+    this.documents.onDidClose(event => {
+      this.cleanPendingValidation(event.document);
+      this.cleanDiagnostics(event.document);
+    });
+
+    this.connection.onInitialize(params => {
+      if (params.rootPath) {
+        this.workspaceRoot = URI.file(params.rootPath);
+      } else if (params.rootUri) {
+        this.workspaceRoot = URI.parse(params.rootUri);
+      }
+      this.connection.console.log('The server is initialized.');
+      return {
+        capabilities: {
+          textDocumentSync: this.documents.syncKind,
+          codeActionProvider: false,
+          completionProvider: {
+            resolveProvider: true,
+          },
+          hoverProvider: true,
+          foldingRangeProvider: false,
+        },
+      };
+    });
+    this.connection.onCompletion(params => this.completion(params));
+    this.connection.onHover(params => this.hover(params));
+
+    this.connection.onRequest((method, params) => {
+      if (InitializeDocumentsMethodName === method) {
+        const { uri, content, template } = params;
+        this.LGDocuments.push({ uri, content, template });
+        // run diagnostic
+        const textDocument = this.documents.get(uri);
+        if (textDocument) {
+          setTimeout(() => {
+            this.validate(textDocument);
+          }, 500);
+        }
+      }
+    });
+  }
+
+  start() {
+    this.connection.listen();
+  }
+
+  protected getLGDocument(document: TextDocument): LGDocument | undefined {
+    const LGDocument = this.LGDocuments.find(item => item.uri === document.uri);
```

Comment:
```suggestion
    const LGDocument = this.LGDocuments.find(({ uri }) => uri === document.uri);
```

Also cache the `document.uri` before calling `find`.

Reply:
changed,
I want ask，it's both safe, but `{ uri }` is better code style  ,right?